scanner/src/main/kotlin/ScannerResultBuilder.kt

```diff
@@ -121,3 +126,79 @@ class InMemoryScannerResultBuilder : ScannerResultBuilder {
      */
     fun result(): OrtResult = ortResult
 }
+
+/**
+ * An implementation of [ScannerResultBuilder] that uses the Jackson Streaming API to construct result files.
+ *
+ * Whenever new data becomes available, it is directly written to the output file. So the memory consumed by this
+ * builder class is reduced.
+ */
+class StreamingScannerResultBuilder(
+    /** The file to write the result to. */
+    val outputFile: File
+) : ScannerResultBuilder {
+    /** Store a reference to the object mapper. */
+    private val mapper = outputFile.mapper()
+
+    /** The generator to produce the result in a streaming fashion. */
+    private val generator = mapper.createGenerator(outputFile.safeMkdirsParent(), JsonEncoding.UTF8)
+
+    /** Stores the labels from the analyzer result. */
+    private var analyzerLabels: Map<String, String>? = null
+
+    override fun initFromAnalyzerResult(analyzerResult: OrtResult) {
+        analyzerLabels = analyzerResult.labels
+
+        with(generator) {
+            writeStartObject()
+            writeObjectField("repository", analyzerResult.repository)
+            writeObjectField("analyzer", analyzerResult.analyzer)
+            writeObjectField("advisor", analyzerResult.advisor)
+            writeObjectField("evaluator", analyzerResult.evaluator)
+
+            writeObjectFieldStart("scanner")
+            writeObjectFieldStart("results")
+            writeArrayFieldStart("scan_results")
+        }
+    }
+
+    override fun addScanResult(resultContainer: ScanResultContainer) {
+        mapper.writeValue(generator, resultContainer)
+    }
+
+    override fun complete(
+        startTime: Instant,
+        endTime: Instant,
+        environment: Environment,
+        config: ScannerConfiguration,
+        storageStats: AccessStatistics,
+        labels: Map<String, String>
+    ) {
+        val allLabels = analyzerLabels.orEmpty() + labels
+
+        with(generator) {
+            writeEndArray()
+            writeObjectField("storage_stats", storageStats)
+            writeEndObject()
+            writeObjectField("start_time", startTime)
+            writeObjectField("end_time", endTime)
+            writeObjectField("environment", environment)
+            writeObjectField("config", config)
+            writeEndObject()
+            writeObjectField("labels", allLabels)
+            writeEndObject()
+        }
```

Comment:
Ok, this explains now the design decisions I comment on in the previous commits. But writing the ORT result immediately to the disk might solve memory issues in the scanner command, but the later steps like the evaluator or reporter will still read the whole result to memory, so if the scan results are the issue this would solve the issue only for the scanner command. I had actually two other problems in mind:
* The stack traces you posted on Slack all pointed to `LocalScanner.getVersion()`. Until a recent change from @sschuberth each call spawned a new ScanCode process. It would be interesting to know if this change already helped.
* The scanner does currently download the raw results from the storage backend, only the throw them away. This makes the download slower and also uses a lot more memory, because the raw results can be significantly larger the scan summaries from the ORT model. I discussed with @sschuberth already that I will remove the raw results next week.

As this change is quite big and also requires API changes, I would appreciate if we could put in hold until the raw results are removed to verify if this already solves the issue completely. Otherwise we should make a plan how the memory issue could also be solved for the evaluator and reporter.

@oheger-bosch  Feel free to schedule a call with @sschuberth,  @fviernau  and me if you would like to discuss the details how we continue with fixing the memory issues.

Reply:
> Until a recent change from @sschuberth each call spawned a new ScanCode process. It would be interesting to know if this change already helped.

According to the replies to https://github.com/oss-review-toolkit/ort/issues/3091#issuecomment-736413602 that change did indeed already help.